### PR TITLE
Storing user information as part of the job when security plugin is installed

### PR DESCRIPTION
**Pull request by thalurur**
*Issue #, if available:*
https://github.com/opensearch-project/index-management/issues/75

*Description of changes:*
Storing user object as part of the job when information is available and setting the user roles in threadcontext while running the ISM policy actions to ensure the user who created managed indices have the necessary permissions before running the action

*CheckList:*
[  ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/index-management/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/ManagedIndexRunner.kt [hunk 1]
```diff
@@ -283,7 +291,7 @@ object ManagedIndexRunner :
         }
 
         val state = policy.getStateToExecute(managedIndexMetaData)
-        val action: Action? = state?.getActionToExecute(clusterService, scriptService, client, settings, managedIndexMetaData)
+        val action: Action? = state?.getActionToExecute(clusterService, scriptService, client, settings, managedIndexMetaData.copy(user = managedIndexConfig.user))
```
Comment (bowenlan-amzn) [inline src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/ManagedIndexRunner.kt:294]: What is this for?

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesResponse.kt [hunk 2]
```diff
@@ -72,7 +72,7 @@ class GetPoliciesResponse : ActionResponse, ToXContentObject {
                         .field(_ID, policy.id)
                         .field(_SEQ_NO, policy.seqNo)
                         .field(_PRIMARY_TERM, policy.primaryTerm)
-                        .field(Policy.POLICY_TYPE, policy, XCONTENT_WITHOUT_TYPE)
+                        .field(Policy.POLICY_TYPE, policy, XCONTENT_WITHOUT_TYPE_AND_USER)
```
Comment (bowenlan-amzn) [inline src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesResponse.kt:75]: Supposed to not showing the user object from this request or temporary thing?

Comment (thalurur) [reply src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/getpolicy/GetPoliciesResponse.kt:75]: Not showing it for now, will have to change this once all the APIs are changed

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/ManagedIndexRunner.kt (hunk 1 above)
Comment (thalurur) [reply src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/ManagedIndexRunner.kt:294]: So the User information is available inside the `Step` for Rollup ISM Action the User object needs to be populated from the managed Index

File: src/main/kotlin/org/opensearch/indexmanagement/util/SecurityUtils.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+class SecurityUtils {
+    companion object {
+        const val INTERNAL_REQUEST = "index_management_plugin_internal_user"
+        val DEFAULT_INJECT_ROLES: List<String> = listOf("all_access", "AmazonES_all_access")
+        /**
+         * Helper method to build the user object either from the threadContext or from the requested user.
+         */
+        fun buildUser(threadContext: ThreadContext, requestedUser: User? = null): User? {
+            if (threadContext.getTransient<Boolean>(INTERNAL_REQUEST) != null && threadContext.getTransient<Boolean>(INTERNAL_REQUEST)) {
```
Comment (dbbaughe) [inline src/main/kotlin/org/opensearch/indexmanagement/util/SecurityUtils.kt:26]: I think these getTransients can throw exceptions, should we wrap these in a try catch or does the caller do it?

File: src/main/kotlin/org/opensearch/indexmanagement/opensearchapi/OpenSearchExtensions.kt [hunk 4]
```diff
@@ -199,3 +212,42 @@ fun <T> XContentParser.parseWithType(
...
+    override val key: CoroutineContext.Key<*>
+        get() = Key
+    val injector = InjectSecurity(id, settings, threadContext)
+
+    /**
+     * Before the thread executes the coroutine we want the thread context to contain user roles so they are used when executing the code inside
+     * the coroutine
+     */
+    override fun updateThreadContext(context: CoroutineContext) {
+        logger.info("Setting security context in thread ${Thread.currentThread().name} for job $id")
```
Comment (dbbaughe) [inline src/main/kotlin/org/opensearch/indexmanagement/opensearchapi/OpenSearchExtensions.kt:235]: I think this (updateThreadContext) gets called every time coroutine resumes right? Could be some log spam w/ info log level, especially once thousands of managed indices running and then add in rollups and transforms etc.

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/util/RestHandlerUtils.kt [hunk 5]
```diff
@@ -106,13 +110,25 @@ data class FailedIndex(val name: String, val uuid: String, val reason: String) :
...
+    user: User? = null
 ): XContentBuilder {
-    return XContentFactory.jsonBuilder()
-        .startObject()
-        .startObject(ManagedIndexConfig.MANAGED_INDEX_TYPE)
-        .optionalTimeField(ManagedIndexConfig.LAST_UPDATED_TIME_FIELD, Instant.now())
-        .field(ManagedIndexConfig.CHANGE_POLICY_FIELD, changePolicy)
-        .endObject()
-        .endObject()
+    if (user == null) {
```
Comment (dbbaughe) [inline src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/util/RestHandlerUtils.kt:116]: Seems like a lot of duplicationg below, can you just set val builder = ... and then if (user == null) builder.field(...etc) and return builder?

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt [hunk 6]
```diff
@@ -175,11 +176,13 @@ class TransportIndexPolicyAction @Inject constructor(
         }
 
         private fun putPolicy() {
-            request.policy.copy(schemaVersion = IndexUtils.indexManagementConfigSchemaVersion)
```
Comment (dbbaughe) [inline src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:178]: Hmm seems like you caught a bug here too, what was the impact? Schema version not getting updated?

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt [hunk 7]
```diff
@@ -273,6 +273,7 @@ class ManagedIndexCoordinator(
      */
     @OpenForTesting
     suspend fun sweepClusterChangedEvent(event: ClusterChangedEvent) {
+        logger.info("test:      processing cluster change event")
```
Comment (dbbaughe) [inline src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:276]: Clean up

**Review by dbbaughe — COMMENTED** [review]
I saw we have user saved on policies, managed index configs, and managed index metadata?
Why do we need the user on all three versus just on the policy?

File: src/main/kotlin/org/opensearch/indexmanagement/util/SecurityUtils.kt (hunk 3 above)
Comment (thalurur) [reply src/main/kotlin/org/opensearch/indexmanagement/util/SecurityUtils.kt:26]: let me check, I haven't seen any errors with getTransient, Alerting and other plugins are using the getTransient as part of every transport action with out any wrapper.

**Comment by thalurur** [comment]
> I saw we have user saved on policies, managed index configs, and managed index metadata?
> Why do we need the user on all three versus just on the policy?

We are note persisting user information in managed index metadata, its just part of the instance object that is passed to Step. We are persisting it in Policy and ManagedIndex as top level fields. Also when policy is copied over to ManagedIndex we are not storing the User object inside the ManagedIndex.Policy path.
The user field in ManagedIndex and Policy will be used to filter the data when fetching this for read APIs

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt (hunk 7 above)
Comment (thalurur) [reply src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:276]: sure, somehow missed it

File: src/main/kotlin/org/opensearch/indexmanagement/opensearchapi/OpenSearchExtensions.kt (hunk 4 above)
Comment (thalurur) [reply src/main/kotlin/org/opensearch/indexmanagement/opensearchapi/OpenSearchExtensions.kt:235]: right, I will make this debug log so its not logged in production by default

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/util/RestHandlerUtils.kt (hunk 5 above)
Comment (thalurur) [reply src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/util/RestHandlerUtils.kt:116]: sure, I actually have `optionalUserField` which is what I want to use eventually. Currently didn't want to return the user field as part of the REST API output at the moment, for now will do what you suggested to reduce code duplication

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt (hunk 6 above)
Comment (thalurur) [reply src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:178]: yes, it is a bug. Earlier we weren't setting the schema version to be latest version but to 1 by default and if I set a random number in input for schema_version that is persisted

**Review by dbbaughe — APPROVED** [review]

**Pull request merged**